#!/bin/bash

###############################################################################
#
#  Amplicon DNA workflow
#
#  This workflow script generates frequency tables from raw bacterial
#  16S rRNA and fungal ITS 1 amplicon data.
#
#  AmpProc and dependent scripts are distributed without guarantees that
#  everything will function flawlessly, although I strive to make it work
#  as best as possible.
#
#  Author: Erika Yashiro, Ph.D.
#  Last updated: 26 September 2023
#
#
#  ampproc_config.sh must include the path to all of the dependent software
#  packages, as well as the paths to the reference database that you need
#  for your project.
#
#  For the software packages, you need to put the path to where the
#  executable file is located. Alternatively, if the executable can be
#  accessed without a path (e.g. if you already have the path in your $PATH)
#  then just write the name of the executable file without the full path.
#
#  The individual reference databases are hard-coded into AmpProc. Therefore
#  you need to ensure that the file names are exactly the same as what is
#  indicated in the template of this config files. The only thing that you
#  should adjust is the path to these files.
#  When you download the databases, you need to make sure that any spaces in
#  the file names are replaced by underscores.
#
#  Custom reference databases can be used with AmpProc but you do not need
#  to specify them in the ampproc_config.sh file. You will be asked for the
#  the file and its full path when you run the workflow script ampproc.
#
#
###############################################################################




# Define the location of the sequences folders
SEQPATH="/home/Dopey/sequences/"

# The software commands need to be findable. You can specify the full path to the executable files or just the name of the commands if your $PATH is already set up to point the correct directories where your executable files are located.

# usearch command. You can use either the 32bit or 64bit versions, depending on your how much data you have to process.
USEARCH=/home/Dopey/Software/usearch11/usearch11.0.667_i86linux64

# mafft command
MAFFT=mafft.bat

# FastTreeMP command
FASTTREEMP=FastTreeMP

# newick script
NEWICK=newick

######################
# Reference databases
######################

# Greengenes 97_otus.fasta for prefiltering at 60% ID
GG97REF="/home/Dopey/dabatases/AmpProc_core_dbs/97_otus.fasta"

# $REFDATABASE == 1
# This is the Silva LTP dbs from Usearch
TAXFILEsilvaLTP132="silvaLTP132"
TAXVERSsilvaLTP132="LPT v132"
REFDATAPATHsilvaLTP132="/home/Dopey/dabatases/SILVA/LTPs132_SSU_unaligned.sintax.fasta"
REFNOTEsilvaLTP132="using SILVA LTP reference database."

# $REFDATABASE == 2
# This is the Silva v132 that was generated by the Qiime developers and formatted for sintax by E. Yashiro.
TAXFILEsilva99pc132="silva99pc132"
TAXVERSsilva99pc132="qiime-formatted SSUREF99% v132"
REFDATAPATHsilva99pc132="/home/Dopey/dabatasesSILVA/silva_132_qiime99_16S_sorted.sintax.fasta"
REFNOTEsilva99pc132="using SILVA qiime99% v132 reference database."

# $REFDATABASE == 3
# This is the SILVA v 138.1 SSURef 99% trunc dbs that was formatted for sintax by E. Yashiro.
TAXFILEsilva99pc1381="silva99pc1381"
TAXVERSsilva99pc1381="SILVA SSUREFNR99% trunc v138.1"
REFDATAPATHsilva99pc1381="/home/Dopey/dabatases/SILVA/SILVA_138.1_SSURef_NR99_tax_silva_trunc.sintax.fasta"
REFNOTEsilva99pc1381="using SILVA SSURef NR99 v138.1 16S & 18S reference database."

# $REFDATABASE == 4
# Link the MiDAS v2.1.3 file.
# https://www.midasfieldguide.org/files/downloads/taxonomies/SINTAX%20fa%20file%20MiDAS%202.1.3.fa
TAXFILEmidas2="midas2"
TAXVERSmidas2="v2.1.3"
REFDATAPATHmidas2="/home/Dopey/dabatases/SINTAX_fa_file_MiDAS_2.1.3.fa"
REFNOTEmidas2="using MiDAS v2 reference database."

# $REFDATABASE == 5
# Link the MiDAS v3.7 dbs of the Danish wastewater microbiomes.
# https://www.midasfieldguide.org/files/downloads/taxonomies/SINTAX%20fa%20file%20MiDAS%203.7.fa
TAXFILEmidas3="midas3"
TAXVERSmidas3="$MIDAS3VERS"
REFDATAPATHmidas3="/home/Dopey/dabatases/midas/SINTAX_fa_file_MiDAS_3.7.fa"
REFNOTEmidas3="using MiDAS $MIDAS3VERSABBREV reference database."

# $REFDATABASE == 6
# Link the latest MiDAS global dbs.
# https://www.midasfieldguide.org/files/downloads/taxonomies/SINTAX%20fa%20MiDAS%205.1.fa
TAXFILEmidas4="midas5"
TAXVERSmidas4="$MIDAS4VERS"
REFDATAPATHmidas4="/home/Dopey/dabatases/midas/SINTAX_fa_MiDAS_5.1.fa"
REFNOTEmidas4="using MiDAS $MIDAS4VERSABBREV reference database."

# $REFDATABASE == 7
# This is the old RDP training set v.16. It is here mostly for reference and test purposes, and should not be used for taxonomy mapping of publishable studies.
TAXFILErdp="rdp"
TAXVERSrdp="training set v16"
REFDATAPATHrdp="/home/Dopey/dabatases/RDP_training_set/rdp_16s_v16s_sp_sintax.cleaned.20180103.fa"
REFNOTErdp="using RDP reference database."

# $REFDATABASE == 8
# This is the UNITE fungal reference dbs. Link the latest version available.
TAXFILEuniteFUN="uniteFUN"
TAXVERSuniteFUN="v9.0 (2023-07-25)"
REFDATAPATHuniteFUN="/home/Dopey/dabatases/unite_fungi_9.0/utax_reference_dataset_fungi_25.07.2023.v9.0.corrected.fasta"
REFNOTEuniteFUN="using UNITE fungi reference database."

# $REFDATABASE == 9
# This is the UNITE eukaryote reference dbs. Link the latest version available.
TAXFILEuniteEUK="uniteEUK"
TAXVERSuniteEUK="v9.0 (2023-07-25)"
REFDATAPATHuniteEUK="/home/Dopey/dabatases/unite_euk_9.0/utax_reference_dataset_all_25.07.2023.v9.0.corrected.fasta"
REFNOTEuniteEUK="using UNITE eukaryotes reference database."

# $REFDATABASE == 10
# This is Mitofish dbs that was downloadable from Mitohelper, and reformatted for sintax by E. Yashiro.
TAXFILE12sMitohelper="12sMitohelper"
TAXVERS12sMitohelper="12S mitosh 2021-03"
REFDATAPATH12sMitohelper="/home/Dopey/dabatases/12S/mitosh.12S.Mar2021.sintax.fasta"
REFNOTE12sMitohelper="using Mitofish 12S eukaryotes reference database."

# $REFDATABASE == 11
# The MIDORI Unique and Long dbs
# https://www.reference-midori.info/download/Databases/GenBank257/SINTAX/uniq/MIDORI2_UNIQ_NUC_GB257_srRNA_SINTAX.fasta.gz
TAXFILE12sMIDORIuniq="12sMIDORIuniq"
TAXVERS12sMIDORIuniq="MIDORI Unique 12S GB257 (2023-09)"
REFDATAPATH12sMIDORIuniq="/home/Dopey/dabatases/12S/MIDORI2_UNIQ_NUC_GB257_srRNA_SINTAX.fasta"
REFNOTE12sMIDORIuniq="using MIDORI 12S Unique metazoan reference database."

# $REFDATABASE == 12
# https://www.reference-midori.info/download/Databases/GenBank257/SINTAX/longest/MIDORI2_LONGEST_NUC_GB257_srRNA_SINTAX.fasta.gz
TAXFILE12SMIDORIlong="12SMIDORIlong"
TAXVERS12SMIDORIlong="MIDORI Longest 12S GB257 (2023-09)"
REFDATAPATH12SMIDORIlong="/home/Dopey/dabatases/12S/MIDORI2_LONGEST_NUC_GB257_srRNA_SINTAX.fasta"
REFNOTE12SMIDORIlong="using MIDORI 12S Longest metazoan reference database."

